src/lib/spinel/radio_spinel.cpp

```diff
@@ -2457,5 +2460,59 @@ void RadioSpinel::HandleCompatibilityError(void)
     DieNow(OT_EXIT_RADIO_SPINEL_INCOMPATIBLE);
 }
 
+void RadioSpinel::MetricsTracker::Init(void) { RestoreMetrics(); }
+
+void RadioSpinel::MetricsTracker::RestoreMetrics(void)
+{
+    VerifyOrExit((mCallbacks.mSaveRadioSpinelMetrics != nullptr) && (mCallbacks.mRestoreRadioSpinelMetrics != nullptr));
+
+    if (mCallbacks.mRestoreRadioSpinelMetrics(mMetrics, mCallbacks.mRadioSpinelMetricsContext) != OT_ERROR_NONE)
+    {
+        memset(&mMetrics, 0, sizeof(mMetrics));
+        mCallbacks.mSaveRadioSpinelMetrics(mMetrics, mCallbacks.mRadioSpinelMetricsContext);
+    }
+
+exit:
+    return;
+}
+
+void RadioSpinel::MetricsTracker::SaveMetrics(void)
+{
+    VerifyOrExit(mCallbacks.mSaveRadioSpinelMetrics != nullptr);
+    mCallbacks.mSaveRadioSpinelMetrics(mMetrics, mCallbacks.mRadioSpinelMetricsContext);
+
+exit:
+    return;
+}
+
+void RadioSpinel::MetricsTracker::IncrementCount(MetricType aType)
+{
+    RestoreMetrics();
+    IncrementMetricCount(aType);
+    SaveMetrics();
+}
+
+void RadioSpinel::MetricsTracker::IncrementMetricCount(MetricType aType)
```

Comment:
How about adding/inlining this whole `switch` block directly `IncremenetCount()` method (can then remove the `IncrementMetricCount()`)?